pallets/delegation/src/delegation_hierarchy.rs

```diff
@@ -52,87 +57,158 @@ impl Default for Permissions {
+		/// The flag indicating whether the delegation has been revoked or not.
+		pub revoked: bool,
+	}
+
+	impl<T: Config> DelegationNode<T> {
+		/// Create a new delegation node that is a direct descendent of the
+		/// given root.
+		///
+		/// * root_id: the root node ID this node will be a child of
+		/// * owner: the identifier of the owner of the new delegation, i.e.,
+		///   the new delegate
+		/// * permissions: the permission flags for the operations the delegate
+		///   is allowed to perform
+		pub fn new_root_child(
+			root_id: DelegationNodeIdOf<T>,
+			owner: DelegatorIdOf<T>,
+			permissions: Permissions,
+		) -> Self {
+			DelegationNode {
+				root_id,
+				owner,
+				permissions,
+				revoked: false,
+				parent: None,
+			}
+		}
+
+		/// Creates a new delegation node that is a direct descendent of the
+		/// given node.
+		///
+		/// * root_id: the root node ID this node will be a child of
+		/// * parent - the parent node ID this node will be a child of
+		/// * owner: the identifier of the owner of the new delegation, i.e.,
+		///   the new delegate
+		/// * permissions: the permission flags for the operations the delegate
+		///   is allowed to perform
+		pub fn new_node_child(
+			root_id: DelegationNodeIdOf<T>,
+			parent: DelegationNodeIdOf<T>,
+			owner: DelegatorIdOf<T>,
+			permissions: Permissions,
+		) -> Self {
+			DelegationNode {
+				root_id,
+				parent: Some(parent),
+				owner,
+				permissions,
+				revoked: false,
+			}
+		}
+	}
 }
 
-impl<T: Config> DelegationNode<T> {
-	/// Create a new delegation node that is a direct descendent of the
-	/// given root.
-	///
-	/// * root_id: the root node ID this node will be a child of
-	/// * owner: the identifier of the owner of the new delegation, i.e., the
-	///   new delegate
-	/// * permissions: the permission flags for the operations the delegate is
-	///   allowed to perform
-	pub fn new_root_child(root_id: DelegationNodeIdOf<T>, owner: DelegatorIdOf<T>, permissions: Permissions) -> Self {
-		DelegationNode {
-			root_id,
-			owner,
-			permissions,
-			revoked: false,
-			parent: None,
+// The new types.
+pub(crate) mod v1 {
+	use super::*;
+
+	#[derive(Clone, Debug, Encode, Decode, PartialEq)]
+	pub struct DelegationNode<T: Config> {
+		pub hierarchy_root_id: DelegationNodeIdOf<T>,
+		pub parent: Option<DelegationNodeIdOf<T>>,
+		pub children: BTreeSet<DelegationNodeIdOf<T>>,
+		pub details: DelegationDetails<T>,
+	}
+
+	impl<T: Config> DelegationNode<T> {
+		pub fn new_root_node(id: DelegationNodeIdOf<T>, details: DelegationDetails<T>) -> Self {
+			Self {
+				hierarchy_root_id: id,
+				parent: None,
+				children: BTreeSet::new(),
+				details,
+			}
+		}
+
+		pub fn new_node(
+			hierarchy_root_id: DelegationNodeIdOf<T>,
+			parent: DelegationNodeIdOf<T>,
+			details: DelegationDetails<T>,
+		) -> Self {
+			let mut new_node = Self::new_root_node(hierarchy_root_id, details);
+			new_node.parent = Some(parent);
+
+			new_node
```

Comment:
i would not call the `new_root_node` method here but also just create the struct. This makes a simple think look more complicated. As someone who never saw the code before I'm unsure what `new_root_node` does? Is there some hidden magic happening? Also when we change `new_root_node` to do something special for root nodes, this would happen for all the nodes then.

Reply:
Addressed in https://github.com/KILTprotocol/mashnet-node/commit/cbc29eb171e169aa6cc223b017b57d2d7b4303bd.